File: server/engine/column_detection.py
"""Direct port of src/lib/excel/columnDetection.ts."""

from dataclasses import dataclass

from engine.aliases import AliasSpec
from engine.fuzzy_match import header_alias_score, normalize_header
from engine.models import DetectedColumn

CONFIDENCE_AUTO_ACCEPT = 0.8
CONFIDENCE_TENTATIVE = 0.5


@dataclass
class _ScoredCandidate:
    header: str
    score: float
    inverted: bool


def detect_columns(headers: list[str], alias_map: dict[str, AliasSpec]) -> list[DetectedColumn]:
    field_keys = list(alias_map.keys())
    normalized_headers = [(h, normalize_header(h)) for h in headers]

    candidates_by_field: dict[str, list[_ScoredCandidate]] = {}
    for field in field_keys:
        spec = alias_map[field]
        candidates: list[_ScoredCandidate] = []
        for raw, normalized in normalized_headers:
            if not normalized:
                continue
            best = 0.0
            for alias in spec.get("aliases", []):
                best = max(best, header_alias_score(normalized, normalize_header(alias)))
            best_inverse = 0.0
            for alias in spec.get("inverse_aliases", []):
                best_inverse = max(best_inverse, header_alias_score(normalized, normalize_header(alias)))
            if best_inverse > best:
                candidates.append(_ScoredCandidate(header=raw, score=best_inverse, inverted=True))
            elif best > 0:
                candidates.append(_ScoredCandidate(header=raw, score=best, inverted=False))
        candidates.sort(key=lambda c: c.score, reverse=True)
        candidates_by_field[field] = candidates

    used_headers: set[str] = set()
    assigned: dict[str, _ScoredCandidate] = {}
    remaining_fields = set(field_keys)

    while remaining_fields:
        best_field: str | None = None
        best_candidate: _ScoredCandidate | None = None
        for field in remaining_fields:
            candidate_list = candidates_by_field.get(field, [])
            top = next((c for c in candidate_list if c.header not in used_headers), None)
            if top and (best_candidate is None or top.score > best_candidate.score):
                best_field = field
                best_candidate = top
        if best_field is None or best_candidate is None:
            break
        assigned[best_field] = best_candidate
        used_headers.add(best_candidate.header)
        remaining_fields.discard(best_field)

    result: list[DetectedColumn] = []
    for field in field_keys:
        candidate = assigned.get(field)
        result.append(
            DetectedColumn(
                field=field,
                matched_header=candidate.header if candidate else None,
                confidence=candidate.score if candidate else 0.0,
                inverted=candidate.inverted if candidate else False,
            )
        )
    return result


def confidence_level(score: float) -> str:
    if score >= CONFIDENCE_AUTO_ACCEPT:
        return "high"
    if score >= CONFIDENCE_TENTATIVE:
        return "medium"
    return "low"

File: src/lib/excel/columnDetection.ts
import { headerAliasScore, normalizeHeader } from '@/lib/fuzzyMatch';
import type { DetectedColumn } from '@/types/columnMapping';
import type { AliasSpec } from '@/lib/excel/aliases';

export const CONFIDENCE_AUTO_ACCEPT = 0.8;
export const CONFIDENCE_TENTATIVE = 0.5;

interface ScoredCandidate {
  header: string;
  score: number;
  inverted: boolean;
}

/**
 * Greedily assigns the best-scoring header to each field, ensuring no header
 * is reused across two fields. Returns one DetectedColumn per field key.
 */
export function detectColumns<K extends string>(
  headers: string[],
  aliasMap: Record<K, AliasSpec>
): DetectedColumn<K>[] {
  const fieldKeys = Object.keys(aliasMap) as K[];
  const normalizedHeaders = headers.map((h) => ({ raw: h, normalized: normalizeHeader(h) }));

  const candidatesByField = new Map<K, ScoredCandidate[]>();

  for (const field of fieldKeys) {
    const spec = aliasMap[field];
    const candidates: ScoredCandidate[] = [];
    for (const { raw, normalized } of normalizedHeaders) {
      if (!normalized) continue;
      let best = 0;
      for (const alias of spec.aliases) {
        best = Math.max(best, headerAliasScore(normalized, normalizeHeader(alias)));
      }
      let bestInverse = 0;
      for (const alias of spec.inverseAliases ?? []) {
        bestInverse = Math.max(bestInverse, headerAliasScore(normalized, normalizeHeader(alias)));
      }
      if (bestInverse > best) {
        candidates.push({ header: raw, score: bestInverse, inverted: true });
      } else if (best > 0) {
        candidates.push({ header: raw, score: best, inverted: false });
      }
    }
    candidates.sort((a, b) => b.score - a.score);
    candidatesByField.set(field, candidates);
  }

  // Greedy global assignment: repeatedly pick the highest remaining (field, header) score.
  const usedHeaders = new Set<string>();
  const assigned = new Map<K, ScoredCandidate>();
  const remainingFields = new Set(fieldKeys);

  while (remainingFields.size > 0) {
    let bestField: K | null = null;
    let bestCandidate: ScoredCandidate | null = null;

    for (const field of remainingFields) {
      const list = candidatesByField.get(field) ?? [];
      const top = list.find((c) => !usedHeaders.has(c.header));
      if (top && (!bestCandidate || top.score > bestCandidate.score)) {
        bestField = field;
        bestCandidate = top;
      }
    }

    if (!bestField || !bestCandidate) break;
    assigned.set(bestField, bestCandidate);
    usedHeaders.add(bestCandidate.header);
    remainingFields.delete(bestField);
  }

  return fieldKeys.map((field) => {
    const candidate = assigned.get(field);
    return {
      field,
      matchedHeader: candidate ? candidate.header : null,
      confidence: candidate ? candidate.score : 0,
      inverted: candidate?.inverted ?? false,
    };
  });
}

export function confidenceLevel(score: number): 'high' | 'medium' | 'low' {
  if (score >= CONFIDENCE_AUTO_ACCEPT) return 'high';
  if (score >= CONFIDENCE_TENTATIVE) return 'medium';
  return 'low';
}
